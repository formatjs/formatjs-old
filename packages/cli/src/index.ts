export {
  default as extractAndWrite,
  extract,
  ExtractCLIOptions,
} from './extract';
