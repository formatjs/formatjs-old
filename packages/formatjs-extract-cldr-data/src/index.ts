/*
 * Copyright 2015, Yahoo Inc.
 * Copyrights licensed under the New BSD License.
 * See the accompanying LICENSE file for terms.
 */
import extractRelativeFields, {
  getAllLocales as getAllDateFieldsLocales,
} from './extract-relative';
import extractUnits, {
  getAllLocales as getAllUnitsLocales,
} from './extract-units';
import extractListPatterns, {
  getAllLocales as getAllListLocales,
} from './extract-list';
import extractCurrencies, {
  getAllLocales as getAllCurrenciesLocales,
} from './extract-currencies';

export interface Opts {
  locales?: string[];
}

export function extractAllRelativeFields(options: Opts = {}) {
  // Default to all CLDR locales if none have been provided.
  const locales = options.locales || getAllDateFieldsLocales();
  return extractRelativeFields(locales);
}

export function extractAllUnits(options: Opts = {}) {
  // Default to all CLDR locales if none have been provided.
  const locales = options.locales || getAllUnitsLocales();
  return extractUnits(locales);
}

export function extractAllListPatterns(options: Opts = {}) {
  // Default to all CLDR locales if none have been provided.
  const locales = options.locales || getAllListLocales();
  return extractListPatterns(locales);
}

export function extractAllCurrencies(options: Opts = {}) {
  // Default to all CLDR locales if none have been provided.
  const locales = options.locales || getAllCurrenciesLocales();
  return extractCurrencies(locales);
}

export {getAllLanguages} from './locales';

export const processAliases = process;
export {getAllLocales as getAllDateFieldsLocales} from './extract-relative';
export {getAllLocales as getAllListLocales} from './extract-list';
export {getAllLocales as getAllUnitsLocales} from './extract-units';
export {getAllLocales as getAllCurrenciesLocales} from './extract-currencies';
