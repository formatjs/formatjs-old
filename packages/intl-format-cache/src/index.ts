/*
Copyright (c) 2014, Yahoo! Inc. All rights reserved.
Copyrights licensed under the New BSD License.
See the accompanying LICENSE file for terms.
*/

// -----------------------------------------------------------------------------

// TODO: Remove this once Intl.RelativeTimeFormat is no longer a draft
interface RelativeTimeFormat {
  format(value: number, unit: string): string;
  formatToParts(value: number, unit: string): { value: string }[];
  resolvedOptions(): ResolvedRelativeTimeFormatOptions;
}
interface ResolvedRelativeTimeFormatOptions
  extends Pick<RelativeTimeFormatOptions, 'numeric' | 'style'> {
  locale: string;
}
interface RelativeTimeFormatOptions {
  localeMatcher: 'best fit' | 'lookup';
  numeric: 'always' | 'auto';
  style: 'long' | 'short' | 'narrow';
}
let RelativeTimeFormat: {
  new (
    locales?: string | string[],
    opts?: RelativeTimeFormatOptions
  ): RelativeTimeFormat;
  (
    locales?: string | string[],
    opts?: RelativeTimeFormatOptions
  ): RelativeTimeFormat;
  supportedLocalesOf(
    locales: string | string[],
    opts?: Pick<RelativeTimeFormatOptions, 'localeMatcher'>
  ): string[];
};

// -- Utilities ----------------------------------------------------------------

function getCacheId(inputs: any[]) {
  return JSON.stringify(
    inputs.map(input =>
      input && typeof input === 'object' ? orderedProps(input) : input
    )
  );
}

function orderedProps(obj: object) {
  return Object.keys(obj)
    .sort()
    .map(k => ({ [k]: obj[k] }));
}

interface MemoizeFormatConstructorFn {
  (constructor: typeof Intl.NumberFormat): (
    ...args: ConstructorParameters<typeof Intl.NumberFormat>
  ) => Intl.NumberFormat;
  (constructor: typeof Intl.DateTimeFormat): (
    ...args: ConstructorParameters<typeof Intl.DateTimeFormat>
  ) => Intl.DateTimeFormat;
  (constructor: typeof RelativeTimeFormat): (
    ...args: ConstructorParameters<typeof RelativeTimeFormat>
  ) => RelativeTimeFormat;
  (constructor: any): (...args: any[]) => any;
}

const memoizeFormatConstructor: MemoizeFormatConstructorFn = (
  FormatConstructor,
  cache = {}
) => (...args) => {
  const cacheId = getCacheId(args);
  let format = cacheId && cache[cacheId];
  if (!format) {
    format = new (FormatConstructor as any)(...args);
    if (cacheId) {
      cache[cacheId] = format;
    }
  }

  return format;
};

export default memoizeFormatConstructor;
