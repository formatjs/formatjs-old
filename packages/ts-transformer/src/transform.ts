import * as ts from 'typescript';
import {MessageDescriptor} from './types';
import {interpolateName} from 'loader-utils';

export type Extractor = (filePath: string, msgs: MessageDescriptor[]) => void;

export type InterpolateNameFn = (
  id?: string,
  defaultMessage?: string,
  description?: string
) => string;

export interface Opts {
  /**
   * Whether the metadata about the location of the message in the source file
   * should be extracted. If `true`, then `file`, `start`, and `end`
   * fields will exist for each extracted message descriptors.
   * Defaults to `false`.
   */
  extractSourceLocation?: boolean;
  /**
   * The ES6 module source name of the React Intl package. Defaults to: `"react-intl"`,
   * but can be changed to another name/path to React Intl.
   *
   * @type {string}
   * @memberof Opts
   */
  moduleSourceName?: string;
  /**
   * Remove `defaultMessage` field in generated js after extraction.
   */
  removeDefaultMessage?: boolean;
  /**
   * Additional component names to extract messages from,
   * e.g: `['FormattedFooBarMessage']`. **NOTE**: By default we check
   * for the fact that `FormattedMessage` is
   * imported from `moduleSourceName` to make sure variable alias works.
   * This option does not do that so it's less safe.
   */
  additionalComponentNames?: string[];
  /**
   * Opt-in to extract from `intl.formatMessage` call with the same restrictions,
   * e.g: has to be called with object literal such as
   * `intl.formatMessage({ id: 'foo', defaultMessage: 'bar', description: 'baz'})`
   */
  extractFromFormatMessageCall?: boolean;
  /**
   * Callback function that gets called everytime we encountered something
   * that looks like a MessageDescriptor
   *
   * @type {Extractor}
   * @memberof Opts
   */
  onMsgExtracted?: Extractor;
  /**
   * webpack-style name interpolation
   *
   * @type {(InterpolateNameFn | string)}
   * @memberof Opts
   */
  overrideIdFn?: InterpolateNameFn | string;
  /**
   * TS Compiler program
   */
  program: ts.Program;
}

const DEFAULT_OPTS: Omit<Opts, 'program'> = {
  onMsgExtracted: () => undefined,
};

function getImportSpecifier(program: ts.Program, node: ts.Node) {
  const symbol = program.getTypeChecker().getSymbolAtLocation(node);
  if (!symbol) {
    return;
  }
  const decls = symbol.getDeclarations();
  if (!Array.isArray(decls) || decls.length !== 1) {
    return;
  }
  return decls[0] as ts.ImportSpecifier;
}

function referenceImport(
  moduleSourceName: string,
  importSpecifier?: ts.ImportSpecifier
): boolean {
  return (
    !!importSpecifier &&
    ts.isNamedImports(importSpecifier.parent) &&
    ts.isImportClause(importSpecifier.parent.parent) &&
    ts.isImportDeclaration(importSpecifier.parent.parent.parent) &&
    (importSpecifier.parent.parent.parent.moduleSpecifier as ts.StringLiteral)
      .text === moduleSourceName
  );
}

function isMultipleMessageDecl(
  node: ts.CallExpression,
  program: ts.Program,
  moduleSourceName: string
) {
  const importSpecifier = getImportSpecifier(program, node.expression);
  if (!importSpecifier) {
    return false;
  }
  return (
    referenceImport(moduleSourceName, importSpecifier) &&
    importSpecifier.name.text === 'defineMessages'
  );
}

function isSingularMessageDecl(
  node: ts.CallExpression | ts.JsxOpeningElement | ts.JsxSelfClosingElement,
  program: ts.Program,
  moduleSourceName: string,
  additionalComponentNames: string[]
) {
  const importSpecifier = getImportSpecifier(
    program,
    ts.isCallExpression(node) ? node.expression : node.tagName
  );
  if (!importSpecifier) {
    return false;
  }
  const compNames = new Set([
    '_',
    'FormattedMessage',
    ...additionalComponentNames,
  ]);
  return (
    referenceImport(moduleSourceName, importSpecifier) &&
    compNames.has(importSpecifier.name.text)
  );
}

function extractMessageDescriptor(
  node:
    | ts.ObjectLiteralExpression
    | ts.JsxOpeningElement
    | ts.JsxSelfClosingElement,
  sf: ts.SourceFile,
  {overrideIdFn, extractSourceLocation}: Opts
): MessageDescriptor | undefined {
  let properties: ts.NodeArray<ts.ObjectLiteralElement> | undefined = undefined;
  if (ts.isObjectLiteralExpression(node)) {
    properties = node.properties;
  } else if (ts.isJsxOpeningElement(node) || ts.isJsxSelfClosingElement(node)) {
    properties = node.attributes.properties;
  }
  const msg: MessageDescriptor = {};
  if (!properties) {
    return;
  }
  properties.forEach(prop => {
    const {name} = prop;
    const initializer =
      ts.isPropertyAssignment(prop) || ts.isJsxAttribute(prop)
        ? prop.initializer
        : undefined;
    if (!initializer || !ts.isStringLiteral(initializer) || !name) {
      return;
    }
    switch (name.getText(sf)) {
      case 'id':
        msg.id = initializer.text;
        break;
      case 'defaultMessage':
        msg.defaultMessage = initializer.text;
        break;
      case 'description':
        msg.description = initializer.text;
        break;
    }
  });
  // We extracted nothing
  if (!Object.keys(msg).length) {
    return;
  }
  if (!msg.id && overrideIdFn) {
    switch (typeof overrideIdFn) {
      case 'string':
        msg.id = interpolateName(
          {sourcePath: sf.fileName} as any,
          overrideIdFn,
          {content: msg.description ? `${msg.defaultMessage}#${msg.description}` : msg.defaultMessage}
        );
        break;
      case 'function':
        msg.id = overrideIdFn(msg.id, msg.defaultMessage, msg.description);
        break;
    }
  }
  if (extractSourceLocation) {
    return {
      ...msg,
      file: sf.fileName,
      start: node.getStart(sf),
      end: node.getEnd(),
    };
  }
  return msg;
}

function isIntlFormatMessageCall(node: ts.CallExpression, sf: ts.SourceFile) {
  const method = node.expression;
  if (!ts.isPropertyAccessExpression(method)) {
    return false;
  }

  return (
    (method.name.getText(sf) === 'formatMessage' &&
      ts.isIdentifier(method.expression) &&
      method.expression.getText(sf) === 'intl') ||
    (ts.isPropertyAccessExpression(method.expression) &&
      method.expression.name.getText(sf) === 'intl')
  );
}

function extractMessageFromJsxComponent(
  node: ts.JsxOpeningElement | ts.JsxSelfClosingElement,
  program: ts.Program,
  sf: ts.SourceFile,
  opts: Opts
): typeof node | undefined {
  const {moduleSourceName = 'react-intl', onMsgExtracted} = opts;
  if (
    !isSingularMessageDecl(
      node,
      program,
      moduleSourceName,
      opts.additionalComponentNames || []
    )
  ) {
    return;
  }
  const msg = extractMessageDescriptor(node, sf, opts);
  if (!msg) {
    return;
  }
  if (typeof onMsgExtracted === 'function') {
    onMsgExtracted(sf.fileName, [msg]);
  }

  const clonedEl = ts.getMutableClone(node);
  clonedEl.attributes = setAttributesInObject(clonedEl.attributes, {
    defaultMessage: opts.removeDefaultMessage ? undefined : msg.defaultMessage,
    id: msg.id,
  }) as ts.JsxAttributes;
  return clonedEl;
}

function setAttributesInObject(
  node: ts.ObjectLiteralExpression | ts.JsxAttributes,
  msg: MessageDescriptor
) {
  const newNode = ts.getMutableClone(node);
  newNode.properties = ts.createNodeArray(
    (['id', 'description', 'defaultMessage'] as Array<keyof MessageDescriptor>)
      .filter(k => !!msg[k])
      .map(k => {
        const val = msg[k];
        const keyNode = ts.createIdentifier(k);
        const valNode = ts.createStringLiteral(val + '');
        if (ts.isJsxAttributes(node)) {
          return ts.createJsxAttribute(keyNode, valNode);
        }
        return ts.createPropertyAssignment(keyNode, valNode);
      })
  ) as ts.NodeArray<ts.ObjectLiteralElementLike>;
  return newNode;
}

function extractMessagesFromCallExpression(
  node: ts.CallExpression,
  program: ts.Program,
  sf: ts.SourceFile,
  opts: Opts
): typeof node | undefined {
  const {moduleSourceName = 'react-intl', onMsgExtracted} = opts;
  if (isMultipleMessageDecl(node, program, moduleSourceName)) {
    const [descriptorsObj, ...restArgs] = node.arguments;
    if (ts.isObjectLiteralExpression(descriptorsObj)) {
      const properties = descriptorsObj.properties as ts.NodeArray<
        ts.PropertyAssignment
      >;
      const msgs = properties
        .map(prop =>
          extractMessageDescriptor(
            prop.initializer as ts.ObjectLiteralExpression,
            sf,
            opts
          )
        )
        .filter((msg): msg is MessageDescriptor => !!msg);
      if (!msgs.length) {
        return;
      }
      if (typeof onMsgExtracted === 'function') {
        onMsgExtracted(sf.fileName, msgs);
      }

      const newNode = ts.getMutableClone(node);
      const clonedDescriptorsObj = ts.getMutableClone(descriptorsObj);
      const clonedProperties = ts.createNodeArray(
        properties.map((prop, i) => {
          if (!ts.isObjectLiteralExpression(prop.initializer)) {
            return prop;
          }
          const clonedNode = ts.getMutableClone(prop);
          clonedNode.initializer = setAttributesInObject(prop.initializer, {
            defaultMessage: opts.removeDefaultMessage
              ? undefined
              : msgs[i].defaultMessage,
            id: msgs[i] ? msgs[i].id : undefined,
          });
          return clonedNode;
        })
      );
      clonedDescriptorsObj.properties = clonedProperties;
      newNode.arguments = ts.createNodeArray([
        clonedDescriptorsObj,
        ...restArgs,
      ]);
      return newNode;
    }
  } else if (
    isSingularMessageDecl(
      node,
      program,
      moduleSourceName,
      opts.additionalComponentNames || []
    ) ||
    (opts.extractFromFormatMessageCall && isIntlFormatMessageCall(node, sf))
  ) {
    const [descriptorsObj, ...restArgs] = node.arguments;
    if (ts.isObjectLiteralExpression(descriptorsObj)) {
      const msg = extractMessageDescriptor(descriptorsObj, sf, opts);
      if (!msg) {
        return;
      }
      if (typeof onMsgExtracted === 'function') {
        onMsgExtracted(sf.fileName, [msg]);
      }

      const newNode = ts.getMutableClone(node);
      newNode.arguments = ts.createNodeArray([
        setAttributesInObject(descriptorsObj, {
          defaultMessage: opts.removeDefaultMessage
            ? undefined
            : msg.defaultMessage,
          id: msg.id,
        }),
        ...restArgs,
      ]);
      return newNode;
    }
  }
}

export function transform(opts: Opts) {
  opts = {...DEFAULT_OPTS, ...opts};
  const {program} = opts;
  return (ctx: ts.TransformationContext): ts.Transformer<ts.SourceFile> => {
    function getVisitor(sf: ts.SourceFile) {
      const visitor: ts.Visitor = (node: ts.Node): ts.Node => {
        const newNode = ts.isCallExpression(node)
          ? extractMessagesFromCallExpression(node, program, sf, opts)
          : ts.isJsxOpeningElement(node) || ts.isJsxSelfClosingElement(node)
          ? extractMessageFromJsxComponent(node, program, sf, opts)
          : undefined;

        return newNode || ts.visitEachChild(node, visitor, ctx);
      };
      return visitor;
    }

    return (sf: ts.SourceFile) => ts.visitNode(sf, getVisitor(sf));
  };
}
