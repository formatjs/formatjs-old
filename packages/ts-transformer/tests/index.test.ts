import {join} from 'path';
import {Project} from 'ts-morph';
import {transform, Opts, MessageDescriptor} from '../src';
import * as ts from 'typescript';

const FILES_TO_TESTS: Record<string, Partial<Opts>> = {
  additionalComponentNames: {
    additionalComponentNames: ['CustomMessage'],
  },
  defineMessages: {},
  extractFromFormatMessage: {
    extractFromFormatMessageCall: true,
  },
  extractSourceLocation: {
    extractSourceLocation: true,
  },
  formatMessageCall: {},
  FormattedMessage: {},
  inline: {},
  moduleSourceName: {
    moduleSourceName: 'react-i18n',
  },
  overrideIdFn: {
    overrideIdFn: (id, defaultMessage, description) => {
      return `HELLO.${id}.${defaultMessage!.length}.${typeof description}`;
    },
    extractFromFormatMessageCall: true,
  },
  removeDefaultMessage: {
    removeDefaultMessage: true,
  },
  noImport: {
    overrideIdFn: '[hash:base64:5]',
    extractFromFormatMessageCall: true,
  },
  removeDescription: {},
};

const FIXTURES_DIR = join(__dirname, 'fixtures');

describe('emit asserts for', function () {
  const filenames = Object.keys(FILES_TO_TESTS);
  filenames.forEach(function (fn) {
    if (fn === 'extractSourceLocation') {
      it(`[special] ${fn}`, function () {
        const output = compile(
          join(FIXTURES_DIR, `${fn}.tsx`),
          FILES_TO_TESTS[fn]
        );
        // Check code output
        expect(output.code).toMatchSnapshot();
        expect(output.msgs).toHaveLength(1);
        expect(output.msgs[0]).toMatchSnapshot({
          defaultMessage: 'Hello World!',
          id: 'foo.bar.baz',
          start: 155,
          end: 222,
          file: expect.stringContaining('extractSourceLocation.tsx'),
        });
      });
    } else {
      it(fn, function () {
        const output = compile(
          join(FIXTURES_DIR, `${fn}.tsx`),
          FILES_TO_TESTS[fn]
        );
        expect(output).toMatchSnapshot();
      });
    }
  });
});

function compile(filePath: string, options?: Partial<Opts>) {
  const project = new Project({
    compilerOptions: {
      experimentalDecorators: true,
      jsx: ts.JsxEmit.React,
      module: ts.ModuleKind.CommonJS,
      moduleResolution: ts.ModuleResolutionKind.NodeJs,
      noEmitOnError: false,
      noUnusedLocals: true,
      noUnusedParameters: true,
      stripInternal: true,
      declaration: false,
      baseUrl: __dirname,
      target: ts.ScriptTarget.ES2015,
      rootDir: __dirname,
      outDir: join(__dirname, 'output'),
    },
  });
  project.addSourceFileAtPathIfExists(filePath);

  let msgs: MessageDescriptor[] = [];

  const result = project.emitToMemory({
    customTransformers: {
      before: [
        transform({
          overrideIdFn: '[hash:base64:10]',
          onMsgExtracted: (_, extractedMsgs) => {
            msgs = msgs.concat(extractedMsgs);
          },
          program: project.getProgram().compilerObject,
          ...(options || {}),
        }),
      ],
    },
  });
  return {
    msgs,
    code: result.getFiles()[0].text,
  };
}
