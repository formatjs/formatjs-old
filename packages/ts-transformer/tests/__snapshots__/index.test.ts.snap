// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emit asserts for [special] extractSourceLocation 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
class Foo extends react_1.Component {
    render() {
        return react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'foo.bar.baz', defaultMessage: 'Hello World!' });
    }
}
exports.default = Foo;
"
`;

exports[`emit asserts for [special] extractSourceLocation 2`] = `
Object {
  "defaultMessage": "Hello World!",
  "end": 222,
  "file": StringContaining "extractSourceLocation.tsx",
  "id": "foo.bar.baz",
  "start": 155,
}
`;

exports[`emit asserts for FormattedMessage 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
class Foo extends react_1.Component {
    render() {
        return (react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'foo.bar.baz', defaultMessage: 'Hello World!' }));
    }
}
exports.default = Foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": "The default message.",
      "id": "foo.bar.baz",
    },
  ],
}
`;

exports[`emit asserts for additionalComponentNames 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
function CustomMessage() { }
class Foo extends react_1.Component {
    render() {
        return (react_1.default.createElement(CustomMessage, { id: \\"greeting-world\\", description: \\"Greeting to the world\\", defaultMessage: \\"Hello World!\\" }));
    }
}
exports.default = Foo;
",
  "msgs": Array [],
}
`;

exports[`emit asserts for defineMessages 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
const msgs = react_intl_1.defineMessages({
    header: {
        id: \\"foo.bar.baz\\",
        defaultMessage: \\"Hello World!\\"
    },
    content: {
        id: \\"foo.bar.biff\\",
        defaultMessage: \\"Hello Nurse!\\"
    },
    kittens: {
        id: \\"app.home.kittens\\",
        defaultMessage: \\"{count, plural, =0 {\\\\uD83D\\\\uDE2D} one {# kitten} other {# kittens}}\\"
    },
    trailingWhitespace: {
        id: \\"trailing.ws\\",
        defaultMessage: \\"   Some whitespace   \\"
    },
    escaped: {
        id: \\"escaped.apostrophe\\",
        defaultMessage: \\"A quoted value ''{value}'\\"
    },
    quoted: {
        id: \\"escaped.apostrophe\\",
        defaultMessage: \\"What's going on\\"
    },
    newline: {
        id: \\"newline\\",
        defaultMessage: \\"this is     a message\\"
    }
});
class Foo extends react_1.Component {
    render() {
        return (react_1.default.createElement(\\"div\\", null,
            react_1.default.createElement(\\"h1\\", null,
                react_1.default.createElement(react_intl_1.FormattedMessage, Object.assign({}, msgs.header))),
            react_1.default.createElement(\\"p\\", null,
                react_1.default.createElement(react_intl_1.FormattedMessage, Object.assign({}, msgs.content))),
            react_1.default.createElement(\\"p\\", null,
                react_1.default.createElement(react_intl_1.FormattedMessage, Object.assign({}, msgs.kittens)))));
    }
}
exports.default = Foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": "The default message",
      "id": "foo.bar.baz",
    },
    Object {
      "defaultMessage": "Hello Nurse!",
      "description": "Another message",
      "id": "foo.bar.biff",
    },
    Object {
      "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
      "description": "Counts kittens",
      "id": "app.home.kittens",
    },
    Object {
      "defaultMessage": "   Some whitespace   ",
      "description": "Whitespace",
      "id": "trailing.ws",
    },
    Object {
      "defaultMessage": "A quoted value ''{value}'",
      "description": "Escaped apostrophe",
      "id": "escaped.apostrophe",
    },
    Object {
      "defaultMessage": "What's going on",
      "description": "Escaped apostrophe",
      "id": "escaped.apostrophe",
    },
    Object {
      "defaultMessage": "this is     a message",
      "description": "this is     a     description",
      "id": "newline",
    },
  ],
}
`;

exports[`emit asserts for extractFromFormatMessage 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
const objectPointer = {
    id: 'foo.bar.invalid',
    defaultMessage: 'This cannot be extracted',
    description: 'the plugin only supports inline objects',
};
class Foo extends react_1.Component {
    render() {
        const { intl } = this.props;
        const msgs = {
            baz: this.props.intl.formatMessage({
                id: \\"foo.bar.baz\\",
                defaultMessage: \\"Hello World!\\"
            }),
            biff: intl.formatMessage({
                id: \\"foo.bar.biff\\",
                defaultMessage: \\"Hello Nurse!\\"
            }),
            invalid: this.props.intl.formatMessage(objectPointer),
        };
        return (react_1.default.createElement(\\"div\\", null,
            react_1.default.createElement(\\"h1\\", null, msgs.header),
            react_1.default.createElement(\\"p\\", null, msgs.content),
            react_1.default.createElement(\\"span\\", null,
                react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'foo', defaultMessage: 'bar' }))));
    }
}
exports.default = react_intl_1.injectIntl(Foo);
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": "The default message",
      "id": "foo.bar.baz",
    },
    Object {
      "defaultMessage": "Hello Nurse!",
      "description": "Another message",
      "id": "foo.bar.biff",
    },
    Object {
      "defaultMessage": "bar",
      "description": "baz",
      "id": "foo",
    },
  ],
}
`;

exports[`emit asserts for formatMessageCall 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
const objectPointer = {
    id: 'foo.bar.invalid',
    defaultMessage: 'This cannot be extracted',
    description: 'the plugin only supports inline objects',
};
class Foo extends react_1.Component {
    render() {
        const msgs = {
            baz: this.props.intl.formatMessage({
                id: 'foo.bar.baz',
                defaultMessage: 'Hello World!',
                description: 'The default message',
            }),
            biff: this.props.intl.formatMessage({
                id: 'foo.bar.biff',
                defaultMessage: 'Hello Nurse!',
                description: 'Another message',
            }),
            invalid: this.props.intl.formatMessage(objectPointer),
        };
        return (react_1.default.createElement(\\"div\\", null,
            react_1.default.createElement(\\"h1\\", null, msgs.header),
            react_1.default.createElement(\\"p\\", null, msgs.content),
            react_1.default.createElement(\\"span\\", null,
                react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'foo', defaultMessage: 'bar' }))));
    }
}
exports.default = react_intl_1.injectIntl(Foo);
",
  "msgs": Array [
    Object {
      "defaultMessage": "bar",
      "description": "baz",
      "id": "foo",
    },
  ],
}
`;

exports[`emit asserts for inline 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
const macro_1 = require(\\"@formatjs/macro\\");
class Foo extends react_1.Component {
    render() {
        return (react_1.default.createElement(\\"div\\", null,
            react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'foo.bar.baz', defaultMessage: 'Hello World!' }),
            macro_1._({
                id: 'header',
                defaultMessage: 'Hello World!',
                description: 'The default message',
            }),
            macro_1._({
                id: 'header2',
                defaultMessage: 'Hello World!',
                description: 'The default message',
            })));
    }
}
exports.default = Foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": "The default message.",
      "id": "foo.bar.baz",
    },
  ],
}
`;

exports[`emit asserts for moduleSourceName 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_i18n_1 = require(\\"react-i18n\\");
const react_intl_1 = require(\\"react-intl\\");
// These should be ignored because the \`moduleSourceName\` is configured to be:
// \`react-i18n\`.
const msgs = react_intl_1.defineMessages({
    header: {
        id: 'header',
        defaultMessage: 'Hello World!',
        description: 'The default message',
    },
    content: {
        id: 'content',
        defaultMessage: 'Hello Nurse!',
        description: 'Another message',
    },
});
class Foo extends react_1.Component {
    render() {
        return (react_1.default.createElement(\\"div\\", null,
            react_1.default.createElement(react_i18n_1.FormattedMessage, { id: 'foo.bar.baz', defaultMessage: 'Hello World!' }),
            msgs));
    }
}
exports.default = Foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": "The default message.",
      "id": "foo.bar.baz",
    },
  ],
}
`;

exports[`emit asserts for noImport 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
function foo() {
    props.intl.formatMessage({
        id: \\"SRX4p\\",
        defaultMessage: \\"props {intl}\\"
    }, { bar: 'bar' });
    this.props.intl.formatMessage({
        id: \\"12zCC\\",
        defaultMessage: \\"this props {intl}\\"
    }, { bar: 'bar' });
    return intl.formatMessage({
        id: \\"2kzY8\\",
        defaultMessage: \\"foo {bar}\\"
    }, { bar: 'bar' });
}
exports.foo = foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "props {intl}",
      "description": "bar",
      "id": "SRX4p",
    },
    Object {
      "defaultMessage": "this props {intl}",
      "description": "bar",
      "id": "12zCC",
    },
    Object {
      "defaultMessage": "foo {bar}",
      "description": "bar",
      "id": "2kzY8",
    },
  ],
}
`;

exports[`emit asserts for overrideIdFn 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
const msgs = react_intl_1.defineMessages({
    header: {
        id: \\"foo.bar.baz\\",
        defaultMessage: \\"Hello World!\\"
    },
    content: {
        id: \\"foo.bar.biff\\",
        defaultMessage: \\"Hello Nurse!\\"
    }
});
class Foo extends react_1.Component {
    render() {
        this.props.intl.formatMessage({
            id: \\"HELLO.undefined.5.string\\",
            defaultMessage: \\"no-id\\"
        });
        return (react_1.default.createElement(\\"div\\", null,
            react_1.default.createElement(\\"h1\\", null,
                react_1.default.createElement(react_intl_1.FormattedMessage, Object.assign({}, msgs.header))),
            react_1.default.createElement(\\"p\\", null,
                react_1.default.createElement(react_intl_1.FormattedMessage, Object.assign({}, msgs.content))),
            react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'foo.bar.zoo', defaultMessage: 'Hello World!' }),
    }
}
exports.default = Foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": "The default message",
      "id": "foo.bar.baz",
    },
    Object {
      "defaultMessage": "Hello Nurse!",
      "id": "foo.bar.biff",
    },
    Object {
      "defaultMessage": "no-id",
      "description": "no-id",
      "id": "HELLO.undefined.5.string",
    },
    Object {
      "defaultMessage": "Hello World!",
      "id": "foo.bar.zoo",
    },
    Object {
      "defaultMessage": "<h1>Hello World!</h1>",
      "description": "The default message.",
      "id": "foo.bar.delta",
    },
  ],
}
`;

exports[`emit asserts for removeDefaultMessage 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
const messages = react_intl_1.defineMessages({
    foo: {
        id: \\"greeting-user\\"
    }
});
class Foo extends react_1.Component {
    render() {
        return (react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'greeting-world' }));
    }
}
exports.default = Foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello, {name}",
      "description": "Greeting the user",
      "id": "greeting-user",
    },
    Object {
      "defaultMessage": "Hello World!",
      "description": "Greeting to the world",
      "id": "greeting-world",
    },
  ],
}
`;

exports[`emit asserts for removeDescription 1`] = `
Object {
  "code": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const react_1 = require(\\"react\\");
const react_intl_1 = require(\\"react-intl\\");
const messages = react_intl_1.defineMessages({
    foo: {
        id: \\"greeting-user\\",
        defaultMessage: \\"Hello, {name}\\"
    }
});
class Foo extends react_1.Component {
    render() {
        return (react_1.default.createElement(react_intl_1.FormattedMessage, { id: 'greeting-world', defaultMessage: 'Hello World!' }));
    }
}
exports.default = Foo;
",
  "msgs": Array [
    Object {
      "defaultMessage": "Hello, {name}",
      "description": "Greeting the user",
      "id": "greeting-user",
    },
    Object {
      "defaultMessage": "Hello World!",
      "description": "Greeting to the world",
      "id": "greeting-world",
    },
  ],
}
`;
